velox/functions/prestosql/types/IPAddressType.cpp

```diff
@@ -0,0 +1,238 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/prestosql/types/IPAddressType.h"
+#include <folly/IPAddress.h>
+#include "velox/expression/CastExpr.h"
+
+static constexpr int kIPV4AddressBytes = 4;
+static constexpr int kIPV4ToV6FFIndex = 10;
+static constexpr int kIPV4ToV6Index = 12;
+static constexpr int kIPAddressBytes = 16;
+static constexpr int kIPAddressMaxStrLen = 39;
+
+namespace facebook::velox {
+
+namespace {
+
+class IPAddressCastOperator : public exec::CastOperator {
+ public:
+  bool isSupportedFromType(const TypePtr& other) const override {
+    switch (other->kind()) {
+      case TypeKind::VARBINARY:
+      case TypeKind::VARCHAR:
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  bool isSupportedToType(const TypePtr& other) const override {
+    switch (other->kind()) {
+      case TypeKind::VARBINARY:
+      case TypeKind::VARCHAR:
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  void castTo(
+      const BaseVector& input,
+      exec::EvalCtx& context,
+      const SelectivityVector& rows,
+      const TypePtr& resultType,
+      VectorPtr& result) const override {
+    context.ensureWritable(rows, resultType, result);
+
+    if (input.typeKind() == TypeKind::VARCHAR) {
+      castFromString(input, context, rows, *result);
+    } else if (input.typeKind() == TypeKind::VARBINARY) {
+      castFromVarbinary(input, context, rows, *result);
+    } else {
+      VELOX_UNSUPPORTED(
+          "Cast from {} to IPAddress not supported", resultType->toString());
+    }
+  }
+
+  void castFrom(
+      const BaseVector& input,
+      exec::EvalCtx& context,
+      const SelectivityVector& rows,
+      const TypePtr& resultType,
+      VectorPtr& result) const override {
+    context.ensureWritable(rows, resultType, result);
+
+    if (resultType->kind() == TypeKind::VARCHAR) {
+      castToString(input, context, rows, *result);
+    } else if (resultType->kind() == TypeKind::VARBINARY) {
+      castToVarbinary(input, context, rows, *result);
+    } else {
+      VELOX_UNSUPPORTED(
+          "Cast from IPAddress to {} not supported", resultType->toString());
+    }
+  }
+
+ private:
+  static void castToString(
+      const BaseVector& input,
+      exec::EvalCtx& context,
+      const SelectivityVector& rows,
+      BaseVector& result) {
+    auto* flatResult = result.as<FlatVector<StringView>>();
+    const auto* ipaddresses = input.as<SimpleVector<int128_t>>();
+    folly::ByteArray16 addrBytes;
+
+    flatResult->resize(rows.size());
```

Comment:
This is not right. I'm out next week. Maybe @bikramSingh91 can help.

Reply:
@mohsaka Thank you for contributing.
Let me get back to you by EOD today.

Reply:
@bikramSingh91 Thank you. I went off of https://github.com/facebookincubator/velox/blob/c0f0a31fcf67828d9ad4bef23812afb7456651b9/velox/exec/RowContainer.cpp#L676
as stated before. Which when I looked at it it seemed to me to be
rows.size() - number of elements
size_t totalBytes = flagBytes_ * rows.size() + fixedWidthRowSize * rows.size(); - Total number of bytes for all of the elements. 

I guess I misunderstood something. Thank you for helping me out.

Reply:
@mohsaka thanks for the update, I unfortunately did not get time to review this today and need to head out of work, I will try to get to it over the weekend.

Reply:
context: `rows` here is a selectivity vector which can have inactive rows at the end. Therefore you can have cases where rows.size() is say 10 but has 6 inactive trailing rows. Therefore, if you use rows.size() then you would be allocating extra space than which is actually required as in expression eval (as eval here is only concerned with the active rows for efficiency, the code around expression eval ensures those rows are copied over to the final result or this vector is sufficiently resized). rows.end() is therefore more appropriate here which returns one past the last selected value.

(My recommendation) However, to make implementation simpler, we have a utility function `context.ensureWritable()` which takes care of these lower level details for you. See example: https://github.com/facebookincubator/velox/blob/main/velox/expression/CastExpr.cpp#L167

Reply:
`context.ensureWritable()` should already be called in castTo and castFrom. More follow up on
https://github.com/facebookincubator/velox/pull/10596#discussion_r1714195154